asynchbase/src/main/java/com/yahoo/ycsb/db/AsyncHBaseClient.java

```diff
@@ -0,0 +1,410 @@
+    if (fields != null) {
+      scanner.setQualifiers(getQualifierList(fields));
+    }
+    
+    // no filters? *sniff*
+    ArrayList<ArrayList<KeyValue>> rows = null;
+    try {
+      while ((rows = scanner.nextRows().join(joinTimeout)) != null) {
+        for (final ArrayList<KeyValue> row : rows) {
+          final HashMap<String, ByteIterator> rowResult =
+              new HashMap<String, ByteIterator>(row.size());
+          for (final KeyValue column : row) {
+            rowResult.put(new String(column.qualifier()), 
+                // TODO - do we need to clone this array? YCSB may keep it in memory
+                // for a while which would mean the entire KV would hang out and won't
+                // be GC'd.
+                new ByteArrayByteIterator(column.value()));
+            if (debug) {
+              System.out.println("Got scan result for key: " + 
+                  Bytes.pretty(column.key()));
+            }
+          }
+          result.add(rowResult);
+        }
+      }
+      scanner.close().join(joinTimeout);
+      return Status.OK;
+    } catch (InterruptedException e) {
+      System.err.println("Thread interrupted");
+      Thread.currentThread().interrupt();
+    } catch (Exception e) {
+      System.err.println("Failure reading from row with key " + startkey + 
+          ": " + e.getMessage());
+      return Status.ERROR;
+    }
+    
+    return Status.ERROR;
+  }
+
+  @Override
+  public Status update(String table, String key,
+      HashMap<String, ByteIterator> values) {
+    setTable(table);
+    
+    if (debug) {
+      System.out.println("Setting up put for key: " + key);
+    }
+    
+    final byte[][] qualifiers = new byte[values.size()][];
+    final byte[][] byteValues = new byte[values.size()][];
+    
+    int idx = 0;
+    for (final Entry<String, ByteIterator> entry : values.entrySet()) {
+      qualifiers[idx] = entry.getKey().getBytes();
+      byteValues[idx++] = entry.getValue().toArray();
+      if (debug) {
+        System.out.println("Adding field/value " + entry.getKey() + "/"
+            + Bytes.pretty(entry.getValue().toArray()) + " to put request");
+      }
+    }
+    
+    final PutRequest put = new PutRequest(lastTableBytes, key.getBytes(), 
+        columnFamilyBytes, qualifiers, byteValues);
+    if (!durability) {
+      put.setDurable(false);
+    }
+    if (!clientSideBuffering) {
+      put.setBufferable(false);
+      try {
+        client.put(put).join(joinTimeout);
+      } catch (InterruptedException e) {
+        System.err.println("Thread interrupted");
+        Thread.currentThread().interrupt();
+      } catch (Exception e) {
+        System.err.println("Failure reading from row with key " + key + 
+            ": " + e.getMessage());
+        return Status.ERROR;
+      }
+    } else {
+      // hooray! Asynchronous write. But without a callback and an async
+      // YCSB call we don't know whether it succeeded or not
+      client.put(put);
```

Comment:
I set it up to mimic the HBase client so that if ``clientSideBuffering`` is not set, it behaves in a synchronous manner with the ``.join()`` on each operation. Otherwise if they want buffering, it is async and mimics using the ``BufferedMutator``.